opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java

```diff
@@ -59,19 +59,19 @@
     public static final String SYNC_URL = "/rest/event/sync";
     protected static final int EVENT_PULL_LIMIT = 250;
     protected static final int EVENT_PUSH_LIMIT = 50;
-    private static final String ADD_URL = "rest/event/add";
-    private Context context;
-    private HTTPAgent httpAgent;
-    private SyncUtils syncUtils;
-    private Trace eventSyncTrace;
-    private Trace processClientTrace;
-    private String team;
+    public static final String ADD_URL = "rest/event/add";
```

Comment:
Why are we changing the scope for all these methods and variables?

Reply:
We need to override the fetchRetry method to change the requested url based on logic
`public class HnppSyncIntentService extends SyncIntentService {

    protected boolean isEmptyToAdd = false;

    protected synchronized void fetchRetry(final int count, boolean returnCount) {
        try {
            SyncConfiguration configs = CoreLibrary.getInstance().getSyncConfiguration();
            if (configs.getSyncFilterParam() == null || StringUtils.isBlank(configs.getSyncFilterValue())) {
                complete(FetchStatus.fetchedFailed);
                return;
            }

            final ECSyncHelper ecSyncUpdater = ECSyncHelper.getInstance(context);
            String baseUrl = getFormattedBaseUrl();

            Long lastSyncDatetime = ecSyncUpdater.getLastSyncTimeStamp();
            Timber.i("LAST SYNC DT %s", new DateTime(lastSyncDatetime));

            if (httpAgent == null) {
                complete(FetchStatus.fetchedFailed);
                return;
            }

            startEventTrace(FETCH, 0);

            String url = baseUrl + SYNC_URL;
            Response resp;
            if (configs.isSyncUsingPost()) {
                JSONObject syncParams = new JSONObject();
                syncParams.put(configs.getSyncFilterParam().value(), configs.getSyncFilterValue());
                syncParams.put("serverVersion", lastSyncDatetime);
                syncParams.put("limit", getEventPullLimit());
                syncParams.put(AllConstants.RETURN_COUNT, returnCount);
                syncParams.put("isFromAdd",  isEmptyToAdd);
                resp = httpAgent.postWithJsonResponse(url, syncParams.toString());
            } else {
                if(HnppConstants.isPALogin()){
                    ArrayList<String> getVillageList = SSLocationHelper.getInstance().getSelectedVillageId();
                    String vid = "";
                    for(int i = 0; i< getVillageList.size() ; i++){
                        vid = vid + getVillageList.get(i)+",";
                    }
                    if(!vid.isEmpty()){
                        vid = vid.substring(0,vid.length() - 1);
                    }
                    url += "?" + configs.getSyncFilterParam().value() + "=" + configs.getSyncFilterValue() + "&serverVersion=" + lastSyncDatetime + "&limit=" + getEventPullLimit()+"&isEmptyToAdd="+isEmptyToAdd+"&villageIds="+vid;
                    Log.i("URL:PA %s", url);
                }else{
                    url += "?" + configs.getSyncFilterParam().value() + "=" + configs.getSyncFilterValue() + "&serverVersion=" + lastSyncDatetime + "&limit=" + getEventPullLimit()+"&isEmptyToAdd="+isEmptyToAdd;
                    Log.i("URL: %s", url);
                }
                resp = httpAgent.fetch(url);
            }

            if (resp.isUrlError()) {
                FetchStatus.fetchedFailed.setDisplayValue(resp.status().displayValue());
                complete(FetchStatus.fetchedFailed);
                return;
            }

            if (resp.isTimeoutError()) {
                FetchStatus.fetchedFailed.setDisplayValue(resp.status().displayValue());
                complete(FetchStatus.fetchedFailed);
                return;
            }

            if (resp.isFailure() && !resp.isUrlError() && !resp.isTimeoutError()) {
                fetchFailed(count);
                return;
            }

            if (returnCount) {
                totalRecords = resp.getTotalRecords();
            }

            processFetchedEvents(resp, ecSyncUpdater, count);

        } catch (Exception e) {
            Timber.e(e, "Fetch Retry Exception:  %s", e.getMessage());
            fetchFailed(count);
        }
    }
    protected boolean pushECToServer(EventClientRepository db) {
        boolean isSuccessfulPushSync = true;
        isEmptyToAdd = true;
        // push foreign events to server
        int totalEventCount = db.getUnSyncedEventsCount();
        int eventsUploadedCount = 0;


        String baseUrl = CoreLibrary.getInstance().context().configuration().dristhiBaseURL();
        if (baseUrl.endsWith(context.getString(org.smartregister.R.string.url_separator))) {
            baseUrl = baseUrl.substring(0, baseUrl.lastIndexOf(context.getString(org.smartregister.R.string.url_separator)));
        }

        for (int i = 0; i < syncUtils.getNumOfSyncAttempts(); i++) {
            Map<String, Object> pendingEvents = db.getUnSyncedEvents(EVENT_PUSH_LIMIT);

            if (pendingEvents.isEmpty()) {
                break;
            }
            // create request body
            JSONObject request = new JSONObject();
            try {
                if (pendingEvents.containsKey(AllConstants.KEY.CLIENTS)) {
                    Object value = pendingEvents.get(AllConstants.KEY.CLIENTS);
                    request.put(AllConstants.KEY.CLIENTS, value);

                    if (value instanceof List) {
                        eventsUploadedCount += ((List) value).size();
                    }
                }
                if (pendingEvents.containsKey(AllConstants.KEY.EVENTS)) {
                    request.put(AllConstants.KEY.EVENTS, pendingEvents.get(AllConstants.KEY.EVENTS));
                }
            } catch (JSONException e) {
                Timber.e(e);
            }
            isEmptyToAdd = false;
            String jsonPayload = request.toString();
            startEventTrace(PUSH, eventsUploadedCount);
            Response<String> response = httpAgent.post(
                    MessageFormat.format("{0}/{1}",
                            baseUrl,
                            ADD_URL),
                    jsonPayload);
            if (response.isFailure()) {
                Timber.e("Events sync failed.");
                isSuccessfulPushSync = false;
            } else {
                db.markEventsAsSynced(pendingEvents);
                Timber.i("Events synced successfully.");
                stopTrace(eventSyncTrace);
                updateProgress(eventsUploadedCount, totalEventCount);
                break;
            }
        }

        return isSuccessfulPushSync;
    }


}`

Reply:
We need to add a param called "isEmptyToAdd" to track the requested client event need to add. So that this params setting false when the add request goes and finally this value send at "fetchRetry" method.
TO set this variable value we need to override these method as a result we need to change the variables and method identifier to using this.

Reply:
For data syncing we are using two different select query depend on value of  "isEmptyToAdd". When client event send to server for adding data to server(rest/event/add) at that time "rest/event/sync" API get data by simple query, On the other hand when only called rest/event/sync(no call rest/event/add) API for fresh sync  get data by complex and expensive  query. So fresh or data clear sync is quite heavy and this is not used frequently. So just make differentiate of syncing API we are using  "isEmptyToAdd" as param.